### PR TITLE
indirect command encoder

### PR DESCRIPTION
**Pull request by adamnemecek**

### CODE REVIEW

**Comment by adamnemecek** [comment]
@kvark can i haz a review?

File: src/indirect_encoder.rs [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+pub enum MTLIndirectComputeCommand {}
+
+foreign_obj_type! {
+    type CType = MTLIndirectComputeCommand;
+    pub struct IndirectComputeCommand;
+    pub struct IndirectComputeCommandRef;
+}
+
+impl IndirectComputeCommandRef {
+
```
Comment (kvark) [inline src/indirect_encoder.rs]: nit: weird empty line?

File: src/lib.rs [hunk 2]
```diff
@@ -370,26 +370,30 @@ mod sampler;
...
+    heap::*,
+    library::*,
+    pipeline::*,
+    renderpass::*,
+    resource::*,
+    sampler::*,
+    texture::*,
+    types::*,
+    vertexdescriptor::*,
+    indirect_encoder::*,
```
Comment (kvark) [inline src/lib.rs]: let's keep it alphabetically ordered

File: src/indirect_encoder.rs [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+pub enum MTLIndirectCommandBufferDescriptor {}
+
+foreign_obj_type! {
+    type CType = MTLIndirectCommandBufferDescriptor;
+    pub struct IndirectCommandBufferDescriptor;
+    pub struct IndirectCommandBufferDescriptorRef;
+}
+
+impl IndirectCommandBufferDescriptorRef {
+    pub fn command_types(&self) -> MTLIndirectCommandType {
```
Comment (kvark) [inline src/indirect_encoder.rs]: where is the ability to set this field?

File: src/indirect_encoder.rs [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        unsafe {
+            msg_send![self,
+                commandTypes
+            ]
+        }
+    }
+
+    pub fn inherit_buffer(&self) -> bool {
+        unsafe {
```
Comment (kvark) [inline src/indirect_encoder.rs]: let's follow what the other boolean entry points are doing: they get `Bool` and compare it to TRUE explicitly. We can't expect the ObjC message to return a Rust `bool`

File: src/indirect_encoder.rs [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+    }
+}
+
+pub enum MTLIndirectCommandBuffer {}
+
+foreign_obj_type! {
+    type CType = MTLIndirectCommandBuffer;
+    pub struct IndirectCommandBuffer;
+    pub struct IndirectCommandBufferRef;
+    type ParentType = ResourceRef;
```
Comment (kvark) [inline src/indirect_encoder.rs]: I wonder, where does it say in Metal that it derives from `MTLResource`?

Comment (adamnemecek) [reply src/indirect_encoder.rs]: https://developer.apple.com/documentation/metal/mtlindirectcommandbuffer "inherits from" section.

File: src/indirect_encoder.rs (hunk 4 above)
Comment (adamnemecek) [reply src/indirect_encoder.rs]: will do.

File: src/indirect_encoder.rs (hunk 3 above)
Comment (adamnemecek) [reply src/indirect_encoder.rs]: will do.

File: src/lib.rs (hunk 2 above)
Comment (adamnemecek) [reply src/lib.rs]: will do.

File: src/indirect_encoder.rs (hunk 1 above)
Comment (adamnemecek) [reply src/indirect_encoder.rs]: fixed.

File: src/indirect_encoder.rs [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+        unsafe {
+            match msg_send![self, inheritPipelineState] {
+                YES => true,
+                NO => false,
+                _ => unreachable!(),
+            }
+        }
+    }
+
+    pub fn set_inherit_pipeline_state(&self, inherit: bool){
```
Comment (kvark) [inline src/indirect_encoder.rs]: nit: formatting (space between ")" and "{")

Comment (adamnemecek) [reply src/indirect_encoder.rs]: fixed

**Comment by kvark** [comment]
bors retry

**Comment by kvark** [comment]
bors retry

**Pull request merged**